src/rust/engine/fs/store/src/local.rs

```diff
@@ -4,19 +4,394 @@ use super::{EntryType, ShrinkBehavior};
 
 use std::collections::{BinaryHeap, HashSet};
 use std::fmt::Debug;
-use std::io::{self, Read};
-use std::path::Path;
+use std::os::unix::fs::MetadataExt;
+use std::path::{Path, PathBuf};
 use std::sync::Arc;
-use std::time::{self, Duration, Instant};
+use std::time::{Duration, Instant, SystemTime};
 
+use async_trait::async_trait;
 use bytes::Bytes;
-use futures::future;
-use hashing::{Digest, Fingerprint, EMPTY_DIGEST};
-use lmdb::Error::NotFound;
-use lmdb::{self, Cursor, Transaction};
-use sharded_lmdb::{ShardedLmdb, VersionedFingerprint};
+use futures::future::{self, join_all, try_join, try_join_all};
+use hashing::{
+  async_copy_and_hash, async_verified_copy, AgedFingerprint, Digest, Fingerprint, EMPTY_DIGEST,
+};
+use sharded_lmdb::ShardedLmdb;
+use std::os::unix::fs::PermissionsExt;
+use task_executor::Executor;
+use tempfile::NamedTempFile;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
 use workunit_store::ObservationMetric;
 
+/// How big a file must be to be stored as a file on disk.
+// NB: These numbers were chosen after micro-benchmarking the code on one machine at the time of
+// writing. They were chosen using a rough equation from the microbenchmarks that are optimized
+// for somewhere between 2 and 3 uses of the corresponding entry to "break even".
+const LARGE_FILE_SIZE_LIMIT: usize = 512 * 1024;
```

Comment:
This will need updating... but luckily, it should be a much more clear cut equation, with no need to account for usage count.